### PR TITLE
Jstor Fetcher

### LINKED ISSUE #6627: add resolver for jstor
**Is your suggestion for improvement related to a problem? Please describe.**
I am typically frustrated when I want to integrate a paper from jstor in my library. jstor is a key social science/humanities platform; not all papers/journals come with doi.

**Describe the solution you'd like**
add resolver that can resolve jstor stable links, taking such a form https://www.jstor.org/stable/26878966

### PR DESCRIPTION
**Pull request by joethei**
added Fetcher for jstor.org

Fixes #6627

- [X] Change in CHANGELOG.md described (if applicable)
- [X] Tests created for changes (if applicable)
- [X] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [X] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            List<BibEntry> entries = new ArrayList<>();
+
+            Document doc = Jsoup.parse(response);
+
+            List<Element> elements = doc.body().getElementsByClass("cite-this-item");
+            for (Element element : elements) {
+                BibtexParser parser = new BibtexParser(importFormatPreferences, new DummyFileUpdateMonitor());
+                String id = element.attr("href").replace("citation/info/", "");
+                try {
+                    String data = IOUtils.toString(new URL(CITE_HOST + id), Charsets.UTF_8);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java:62]: Just use our UrlDownload class:
```
URLDownload download = new URLDownload(url);
String data = download.asString()
```

File: src/test/java/org/jabref/logic/importer/fetcher/JstorFetcherTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                .withField(StandardField.PUBLISHER, "International Forum of Educational Technology & Society")
+                .withField(StandardField.NUMBER, "2")
+                .withField(StandardField.PAGES, "63--73")
+                .withField(StandardField.VOLUME, "20")
+                .withField(StandardField.URL, "http://www.jstor.org/stable/90002164")
+                .withField(StandardField.YEAR, "2017");
+
+        List<BibEntry> entries = fetcher.performSearch("Test Anxiety Analysis of Chinese College Students in Computer-based Spoken English Test");
+        assertTrue(entries.contains(expected));
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/logic/importer/fetcher/JstorFetcherTest.java:38]:
```suggestion
        assertEquals(Collections.singletonList(expected), entries);
```
Using assertEquals has the advantage that you can see the difference when the test fails, e.g. sometimes the bibdata changes on the publisher site.

File: src/main/java/org/jabref/logic/importer/WebFetchers.java [hunk 3]
```diff
@@ -42,6 +36,13 @@
 import org.jabref.model.entry.identifier.DOI;
 import org.jabref.model.entry.identifier.Identifier;
 
+import java.util.Comparator;
+import java.util.HashSet;
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/WebFetchers.java:40]: Something is odd with your import order. Did you import the JabRef code style for Intellij?
https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace#using-jabrefs-code-style

**Review by Siedlerchr — CHANGES_REQUESTED** [review]
Thanks for your contribution! That's a nice addition. It would be cool if you could check if it's possible to extend the fetcher so it's a FullTextFetcher and downloads the pdf?

**Review by Siedlerchr — APPROVED** [review]
LGTM now. Please fix the remaining check style issues.
For external contributes we require that at least two JabRef devs do a code review before it's merged.

**Comment by koppor** [comment]
Thank you for the PR and welcome to the open source world of JabRef. Maybe you already registered for [hacktoberfest](https://hacktoberfest.digitalocean.com/). This PR will definitively count towards Hacktoberfest and we are looking for more.

I will go through your code soon. Meanwhile, may I ask to fix the checktyle issues?

>  Error: eckstyle] [ERROR] /home/runner/work/jabref/jabref/src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java:84:9: 'if' is not followed by whitespace. [WhitespaceAround]
> Error: eckstyle] [ERROR] /home/runner/work/jabref/jabref/src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java:93:9: 'if' is not followed by whitespace. [WhitespaceAround]

I know that we seem to ignore the red crosses at pull requests, but we are working on getting them green again. Especially "checkstyle" always was taken seriously.

File: src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public URL getURLForQuery(String query) throws URISyntaxException, MalformedURLException {
+        URIBuilder uriBuilder = new URIBuilder(SEARCH_HOST);
+        uriBuilder.addParameter("Query", query);
+        return uriBuilder.build().toURL();
+    }
+
+    @Override
+    public Parser getParser() {
+        return inputStream -> {
+            String response = new BufferedReader(new InputStreamReader(inputStream)).lines().collect(Collectors.joining(OS.NEWLINE));
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java:57]: I think, the modern way is:

```suggestion
            String response = new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);
```

(Source: https://stackoverflow.com/a/32352386/873282)

File: src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            List<BibEntry> entries = new ArrayList<>();
+
+            Document doc = Jsoup.parse(response);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java:60]: Why can't you use the method directly reading from the input stream? https://jsoup.org/apidocs/org/jsoup/Jsoup.html#parse(java.io.InputStream,java.lang.String,java.lang.String)

File: src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+            List<Element> elements = doc.body().getElementsByClass("cite-this-item");
+            for (Element element : elements) {
+                BibtexParser parser = new BibtexParser(importFormatPreferences, new DummyFileUpdateMonitor());
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java:64]: There is no need to instantiate the parser for each parse run. Please instantiate before the loop

File: src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                String id = element.attr("href").replace("citation/info/", "");
+                try {
+                    String data = new URLDownload(CITE_HOST + id).asString();
+                    entries.addAll(parser.parseEntries(data));
+                } catch (IOException e) {
+                    throw new ParseException("could not download data from jstor.org", e);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fetcher/JstorFetcher.java:70]:
```suggestion
                    throw new ParseException("Could not download data from jstor.org", e);
```

File: src/test/java/org/jabref/logic/importer/fetcher/JstorFetcherTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    void searchByTitle() throws Exception {
+        List<BibEntry> entries = fetcher.performSearch("ti: \"Test Anxiety Analysis of Chinese College Students in Computer-based Spoken English Test\"");
+        assertEquals(Collections.singletonList(bibEntry), entries);
+    }
+
+    @Test
+    void fetchPDF() throws IOException, FetcherException {
+        Optional<URL> url = fetcher.findFullText(bibEntry);
+        assertTrue(url.isPresent());
+        assertEquals(new URL("https://www.jstor.org/stable/pdf/90002164.pdf"), url.get());
```
Comment (koppor) [inline src/test/java/org/jabref/logic/importer/fetcher/JstorFetcherTest.java:52]: One can merge these two lines

```suggestion
        assertEquals(Optional.of(new URL("https://www.jstor.org/stable/pdf/90002164.pdf")), url);
```

**Review by koppor — CHANGES_REQUESTED** [review]
Looks good. Some small nitpick comments. Hope, you find time to address them.

**Comment by koppor** [comment]
One more thing - if possible - could you add org.jabref.logic.importer.fetcher.SearchBasedFetcherCapabilityTest? That would help @DominikVoigt to add support for JSTOR for his SLR addition to JabRef. - No need do it in this PR, maybe as follow up PR? #hacktoberfest

**Comment by koppor** [comment]
I get 403 at a test:

```
Test supportsJournalSearch() FAILED

  org.jabref.logic.importer.FetcherException: A network error occurred while fetching from jstor.org/open/search?Query=pt%3A%22Test%22

  Caused by: java.io.IOException: Server returned HTTP response code: 403 for URL: jstor.org/open/search?Query=pt%3A%22Test%22

  Caused by: java.io.IOException: Server returned HTTP response code: 403 for URL: jstor.org/open/search?Query=pt%3A%22Test%22
```

I think, it's a permission issue. Thus, I go ahead with merging!

Thank you for the quick follow up!

**Pull request merged**